### PR TITLE
chore(ts-migration): migrates numeric menu and panel connector, widget, and stories

### PR DESCRIPTION
**Pull request by nunomaduro**
This pull request migrates 2 widgets, 1 connector and 2 stories:

- The numeric menu connector, widget, and stories;
- The panel widget, and stories;

### CODE REVIEW

File: src/widgets/panel/panel.tsx [hunk 1]
```diff
@@ -96,27 +123,25 @@ const renderer = ({
...
-  container: HTMLElement | string;
-  [key: string]: any;
-};
-type PanelWidget = (
-  params?: PanelWidgetParams
-) => (
-  widgetFactory: WidgetFactory<{}, NestedWidgetOptions>
-) => (widgetOptions: NestedWidgetOptions) => Widget;
+export type PanelWidget = (
+  params: PanelWidgetOptions
```
Comment (eunjae-lee) [inline src/widgets/panel/panel.tsx:127]: I guess `PanelWidgetOptions` should be renamed to `PanelWidgetParams`?

File: .storybook/playgrounds/default.ts [hunk 2]
```diff
@@ -45,18 +46,14 @@ function instantSearchPlayground({
     }),
   ]);
 
-  const numericMenu = document.createElement('div');
-  leftPanel.appendChild(numericMenu);
+  const numericMenuElement = document.createElement('div');
```
Comment (francoischalifour) [inline .storybook/playgrounds/default.ts:49]: In other stories we call these `numericMenuContainer`.

File: src/connectors/numeric-menu/__tests__/connectNumericMenu-test.ts [hunk 3]
```diff
@@ -435,26 +472,31 @@ See documentation: https://www.algolia.com/doc/api-reference/widgets/numeric-men
...
-      createURL: () => '#',
-    });
+    widget.render!(
+      createRenderOptions({
+        results: new SearchResults(helper.state, [
+          createSingleSearchResponse(),
+        ]),
+        state: helper.state,
+        helper,
+        createURL: () => '#',
```
Comment (francoischalifour) [inline src/connectors/numeric-menu/__tests__/connectNumericMenu-test.ts:497]: You can remove `createURL` when they're empty. They're covered by `createRenderOptions`.

File: src/widgets/panel/panel.tsx (hunk 1 above)
Comment (nunomaduro) [reply src/widgets/panel/panel.tsx:127]: I am not sure, because for types is always have been WidgetOptions.

Anyways, going to address of all of that together.

Just created this task for that: https://algolia.atlassian.net/browse/IFW-1260?atlOrigin=eyJpIjoiZjQ3ODg0ZDg0YmVmNDEzNmJjZTA1ZTAzYWRhNDEyMTEiLCJwIjoiaiJ9

File: src/widgets/numeric-menu/__tests__/numeric-menu-test.ts [hunk 4]
```diff
@@ -55,23 +66,25 @@ describe('numericMenu()', () => {
...
 
     createURL = () => '#';
-    widget.init({ helper, instantSearchInstance: {} });
+    widget.init!({ helper, instantSearchInstance: {} });
   });
 
   it('calls twice render(<RefinementList props />, container)', () => {
-    widget.render({ state, results, createURL });
-    widget.render({ state, results, createURL });
+    widget.render!({ state, results, createURL });
```
Comment (francoischalifour) [inline src/widgets/numeric-menu/__tests__/numeric-menu-test.ts:86]: Any reasons the `createRenderOptions` and `createInitOptions` are not used below?

Comment (nunomaduro) [reply src/widgets/numeric-menu/__tests__/numeric-menu-test.ts:86]: @francoischalifour Nice catch.

Comment (nunomaduro) [reply src/widgets/numeric-menu/__tests__/numeric-menu-test.ts]: Fixed. Note that the snapshots got updated as we use a templateConfig as object instead of undefined.

File: src/widgets/numeric-menu/numeric-menu.tsx [hunk 5]
```diff
@@ -47,77 +50,92 @@ const renderer = ({
...
+   * CSS class to add to each radio element (when using the default template).
+   */
+  radio?: string | string[];
+};
+
+export type NumericMenuTemplates = {
+  /**
+   * Item template, provided with `label` (the name in the configuration), `isRefined`, `url`, `value` (the setting for the filter) data properties.
+   */
+  item?: Template;
```
Comment (francoischalifour) [inline src/widgets/numeric-menu/numeric-menu.tsx:99]: We can type the template [as described in the documentation](https://www.algolia.com/doc/api-reference/widgets/numeric-menu/js/#widget-param-item) since it's generic.

File: src/connectors/numeric-menu/connectNumericMenu.ts [hunk 6]
```diff
@@ -4,83 +4,88 @@ import {
...
    * Higher bound of the option (<=)
    */
-  start: number;
+  start?: number;
+
   /**
    * Lower bound of the option (>=)
    */
-  end: number;
+  end?: number;
```
Comment (Haroenv) [inline src/connectors/numeric-menu/connectNumericMenu.ts:29]: they  aren't both optional, at least one is required, this means it's more something like:

```ts
NumericMenuConnectorParamsItem = {
  label: string;
  start: string;
  end?: string;} | { label: string; start?: string; end: string }
```

File: src/widgets/panel/__tests__/panel-test.ts [hunk 7]
```diff
@@ -25,7 +25,7 @@ beforeEach(() => {
 describe('Usage', () => {
   test('without arguments does not throw', () => {
     expect(() => {
-      panel();
+      panel({});
```
Comment (Haroenv) [inline src/widgets/panel/__tests__/panel-test.ts:28]: this test is to see that we don't destructure without default, giving an object changes it :)

File: src/widgets/panel/panel.tsx [hunk 8]
```diff
@@ -162,10 +187,8 @@ const panel: PanelWidget = (widgetParams = {} as PanelWidgetParams) => {
     footer: cx(suit({ descendantName: 'footer' }), userCssClasses.footer),
   };
 
-  return (widgetFactory: WidgetFactory<{}, NestedWidgetOptions>) => (
-    widgetOptions = {} as NestedWidgetOptions
-  ): Widget => {
-    const { container } = widgetOptions;
+  return widgetFactory => widgetOptions => {
+    const { container } = widgetOptions || ({} as typeof widgetOptions);
```
Comment (Haroenv) [inline src/widgets/panel/panel.tsx:191]: this typeof trick is really cool

File: src/widgets/numeric-menu/numeric-menu.tsx (hunk 5 above)
Comment (nunomaduro) [reply src/widgets/numeric-menu/numeric-menu.tsx:99]: @francoischalifour Can you rephrase what are you asking for? Sorry, didn't got it.

File: src/connectors/numeric-menu/connectNumericMenu.ts (hunk 6 above)
Comment (nunomaduro) [reply src/connectors/numeric-menu/connectNumericMenu.ts:29]: @Haroenv Not sure if we should go far on that one. As example, on API Clients we don't go that far on query rules.

File: src/widgets/panel/__tests__/panel-test.ts (hunk 7 above)
Comment (nunomaduro) [reply src/widgets/panel/__tests__/panel-test.ts:28]: Fixed, thanks.

File: src/widgets/numeric-menu/numeric-menu.tsx (hunk 5 above)
Comment (Haroenv) [reply src/widgets/numeric-menu/numeric-menu.tsx:99]: something roughly like this?

```
Template<{
  /**
   * the name of the attribute.
   */
  attribute: string;
  /**
   *  the label for the option.
   */
  label: string;
  /**
   * the encoded URL of the bounds object with a {start, end} form. This value can be used verbatim in the webpage and can be read by refine directly. If you want to inspect the value, you can do JSON.parse(window.decodeURI(value)) to get the object.
   */
  value: string;
  /**
   *  whether or not the refinement is selected.
   */
  isRefined: boolean;
  /**
   * the URL with the applied refinement.
   */
  url: string;
  /**
   * the CSS classes provided to the widget.
   */
  cssClasses: object;
}>
```

Comment (nunomaduro) [reply src/widgets/numeric-menu/numeric-menu.tsx:99]: @Haroenv Can you confirm that this type used in this generic, is no used elsewhere?

Comment (nunomaduro) [reply src/widgets/numeric-menu/numeric-menu.tsx]: @Haroenv Can you check the contents now?

**Comment by nunomaduro** [comment]
Going to merge so I can move forward. Fell free to write down if you see anything that you don't agree. I can adjust later
